### ISSUE TITLE
No push notification for RAT result

### ISSUE THREAD

**Issue by Jens07** [corona-warn-app/cwa-app-ios#3231]
## Avoid duplicates
* [X] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [X] Bug is specific for iOS only, for general issues / questions that apply to iOS and Android please raise them in the [documentation repository](https://github.com/corona-warn-app/cwa-documentation)
* [x] Bug is not already reported in another issue

## Technical details

- Device name: iPhone 11
- iOS version: iOS 14.6
- App version: 2.5.1

## Describe the bug
It doesn't appear a push notification for the result of a rapid test after registering the test in the corona-warn-app.

## Steps to reproduce the issue
Register a rapid test via link or qr-code before or after taking the smear.

## Expected behaviour
It appears a push notification on the iPhone after the result is available.

## Possible Fix
Unknown.

## Additional context

----
Internal Tracking-ID: [EXPOSUREAPP-8709](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8709)

**Comment by Ein-Tim** [comment]
@Jens07 

Do you have "Settings" -> "Notifications" -> "Status of your COVID-19 test" active?

**Comment by Jens07** [comment]
> @Jens07 
> 
> Do you have "Settings" -> "Notifications" -> "Status of your COVID-19 test" active? 

Yes. All the time. I never changed that setting. But I never get a notification. I always had to open the result manually in my app.

**Comment by Ein-Tim** [comment]
Okay thanks @Jens07 

Can you check if CWA is allowed to access notifications in the iOS settings? I guess so. 
If yes, please [create an event on the website](https://www.coronawarn.app/en/eventregistration/) (or in the app, how you like, set the automatic check out time to 15 and close the app (you can either just exit it by swiping from the bottom to the top or also kill the app via the app switch.
Either way, make sure that CWA is allowed to use background processes ("Hintergrundaktualisierung").

If you see a notification from CWA that you have been checked out, we know that this is a problem most likely only with test results. If you don't see a notification from CWA, we know that it's a more general problem.

**Comment by Jens07** [comment]
> If you see a notification from CWA that you have been checked out, we know that this is a problem most likely only with test results.

I got the notification on my lock screen.
![image](https://user-images.githubusercontent.com/68737429/126716718-01cc2304-531a-4399-b086-8a520bc9f3bd.jpeg)

**Comment by Ein-Tim** [comment]
@Jens07 

Okay thanks... Strange, to me it seems like the CWA has access to notifications and it should work. 
I'd say this is all I can do for you at the moment, the devs should take a look here.

**Comment by Ein-Tim** [comment]
I registered a PCR-test in the app yesterday and got the notification that my test result is available ("Es gibt Neuigkeiten von Ihrer Corona-Warn-App") today at 10:00h.

So I am **not** able to reproduce this on my iPhone XR, running iOS 14.7.1 and CWA 2.5.1.

**Comment by heinezen** [comment]
@Jens07 

Thanks for reporting the issue. I've created an internal ticket for the dev team. They will try to investigate the problem.

----
Corona-Warn-App Open Source Team

**Comment by Jens07** [comment]
> I registered a PCR-test in the app yesterday and got the notification that my test result is available ("Es gibt Neuigkeiten von Ihrer Corona-Warn-App") today at 10:00h.

Please note that I never registered a PCR-test. In all cases I registered RAPID TESTS.

**Comment by Ein-Tim** [comment]
I can **confirm** this issue. After I had a RAT done at 17:44h, received the mail with the result at 17:04h there was no push notification in the CWA until now. I have opened the app now, since I wanted to check if there is a certificate available for this test. 

This has to be fixed. My app was hard closed. iPhone XR, iOS 14.7.1, CWA 2.6.2.

**Comment by thomasaugsten** [comment]
Because of the local notification we can notify immediately. On iOS i can take up to 2 hours for the notification.

**Comment by Ein-Tim** [comment]
@thomasaugsten 

I also didn't receive a notification after 2h, not even after nearly 3. 

So I'd suggest that you check this internally and make sure everything is alright.

**Comment by thomasaugsten** [comment]
Maybe you can retest this until you receive a notification

**Comment by Ein-Tim** [comment]
@thomasaugsten I will be off for vacation next week but afterwards I can test 👍

**Comment by jucktnich** [comment]
@thomasaugsten can't an App set the importance of its push notification? I thought you can register a notification as time critical, so the OS plays the notification immediately.

**Comment by thomasaugsten** [comment]
It is a little more complex on iOS. At the end iOS will decide how often you receive background time with the ENF we have privilege status but not enough to receive every  20min background time

**Comment by jucktnich** [comment]
Ah yeah now I get the point. You wanted to say, that you can't fetch the test result that often (and push notifications aren't possible cause of privacy concerns), right?

**Comment by jucktnich** [comment]
Would it be ok￼ from a privacy side (and technically feasible) to send an empty silent push notification to *all* CWAs, so they can fetch the data they need.

**Comment by thomasaugsten** [comment]
Silent push notifications involves Apple and this result in privacy concerns

**Comment by jucktnich** [comment]
@thomasaugsten i just want to push a "please wake up and if you need to, fetch your data"

**Comment by Jens07** [comment]
@thomasaugsten 

> Because of the local notification we can notify immediately. On iOS i can take up to 2 hours for the notification.

The result of a RAT is usually available after 15 minutes. I fear that nearly nobody will wait until 2 hours. It's the sense of a RAT to get a result in a short time. The consequence will be that the users assume that there is no notification service or they suppose a failure in the cwa. I really can't imagine that users will wait 2 hours long.

I think it's a good idea of @jucktnich to inform the user with a push notification. It should dissipate the privacy concerns if the only content of the message is that there are "any news" for the user. So nobody finds out if the notification was about a test, what kind of test or actually the result.

If I go to my next RAT, I wont open the cwa and will wait if I get a notification.

**Comment by Jens07** [comment]
@thomasaugsten 
> Maybe you can retest this until you receive a notification

I'm not @Ein-Tim but I also made another trial to get a notification after a RAT. The result was available since 5th August, 4:57 pm on the website of my test center. Until 7:10 pm I didn't get a notification for the CWA. Unfortunately I couldn't test it any longer because I had to open the app for the "check in" function.

**Comment by dsarkar** [comment]
@Jens07 Thanks for the info. forwarded to internal ticket.

**Comment by ndegendogo** [comment]
> Would it be ok￼ from a privacy side (and technically feasible) to send an empty silent push notification to all CWAs, so they can fetch the data they need.

@jucktnich if I understand your proposal right - no, it is not a good idea to wakeup the devices of **all** cwa users with a short frequency like 20 minutes or less, only because one user is waiting for their RAT result. It would have a massive impact on battery time for all with very little benefit.

However, what I could imagine is to setup a local timer; only after the RAT is registered; first timeslot according to the expected processing time of the RAT (10 minutes? 15 minutes?), then in short sequence (every 1 or 2 minutes?), till the result is available; then stop the timer. Maybe even an increasing interval if the result is really late.
And if course this timer should focus absolutely, and only poll the RAT result server **and nothing else** (no data donation / statistic fetch / not even fetch DEK keys).
This strategy would indeed burn battery and background processing time, but at a time where the user is willing to accepts this 
*(and hopefully the same is true for Apple!*)

**Comment by thomasaugsten** [comment]
On iOS you cannot fix schedule background time. The OS decided when you are allowed to get background execution time. You can only trigger an request

**Comment by jucktnich** [comment]
And is it ok, from a privacy POV ok, to send a silent push notification to every device waiting for an result/rat result? Maybe ask the user if it's ok for them. (The server sending the push notifications needs to be independent from the other servers, this server should just take the request send the silent push notification in a given Intervall for this time to this IP-Adress, nothing more).

**Comment by Ein-Tim** [comment]
@thomasaugsten 

But I mean CWA can also send me a push notification at exactly that time I was checked-out from an event so why can't CWA schedule what @ndegendogo proposed?

**Comment by thomasaugsten** [comment]
This is a scheduled notification. The user will receive a notification but you will not grant background time to check for example test results

**Comment by Ein-Tim** [comment]
Okay, thanks. Then maybe it should be investigated if it makes sense to also schedule a notification in 15 min when a RAT is registered. This notification could say: "Please don't forget to open your Corona-Warn-App to receive your test result."

**Comment by ndegendogo** [comment]
> And is it ok, from a privacy POV ok, to send a silent push notification to every device waiting for an result/rat result?

@jucktnich I may be wrong here; but as to my understanding the server has no awareness which **devices** are waiting for a RAT result, and cannot address them selectively.
The whole data flow was specified with the server agnostic, and responsibility of the device to poll.

**Comment by jucktnich** [comment]
@ndegendogo My proposal is, like stated in parentheses, that there's an independent server, which the app calls, ￼and then the server pushes silent notifications every 20mins for 3 days to the calling IP-Address.

**Comment by Jens07** [comment]
@thomasaugsten 
> Maybe you can retest this until you receive a notification

Today, 11th August, I made another RAT. The result was, according to the website, available since 6:53 pm. At 7:29 pm I got the push notification from the cwa. I used, just as my test on 5th August, iOS 14.7.1 and CWA-Version 2.6.3.

**Comment by Ein-Tim** [comment]
I did a rapid antigen test today. 

Test booked: 11:30 
Test registered in CWA: 11:40
Swap done: 12:35
CWA hard closed at: 12:45
Result via Mail: 13:01
Push Notification from CWA: 15:14

CWA was hard closed (not running in the background). 

Connected to the internet via cellular (LTE) or WLAN the whole time. 
The push notification came in with WLAN connection.

**Comment by svengabr** [comment]
I have just checked the linked Jira issue.

Jira Ticket is flagged as:
Resolution: `Wont Fix`
Status: `Obsolete`

Developer comment:
> I can see in the linked GitHub issue that a retest was successful and they received a notification. It is known that we only check for test results when the background task is running so it can take a little bit longer, up to several hours.
> 
> Apart from that there recently was a fix that improved the reliability of test result notifications as they sometimes could get lost altogether: EXPOSUREAPP-6908
